R/pr.R

```diff
@@ -0,0 +1,556 @@
+#' @param secure Boolean that adds the \code{Secure} cookie flag.  This should be set
+#'   when the route is eventually delivered over \href{https://en.wikipedia.org/wiki/HTTPS}{HTTPS}.
+#' @seealso \itemize{
+#' \item \href{https://github.com/jeroen/sodium}{'sodium'}: R bindings to 'libsodium'
+#' \item \href{https://download.libsodium.org/doc/}{'libsodium'}: A Modern and Easy-to-Use Crypto Library
+#' \item \href{https://github.com/r-lib/keyring}{'keyring'}: Access the system credential store from R
+#' \item \href{https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#Directives}{Set-Cookie flags}: Descriptions of different flags for \code{Set-Cookie}
+#' \item \href{https://developer.mozilla.org/en-US/docs/Glossary/Cross-site_scripting}{Cross-site scripting}: A security exploit which allows an attacker to inject into a website malicious client-side code
+#' }
+#' @examples
+#' \dontrun{
+#'
+#' ## Set secret key using `keyring` (preferred method)
+#' keyring::key_set_with_value("plumber_api", password = plumber::randomCookieKey())
+#'
+#'
+#' pr() %>%
+#'   pr_cookie(
+#'     keyring::key_get("plumber_api"),
+#'     name = "counter"
+#'   ) %>%
+#'   pr_get("/sessionCounter", function(req) {
+#'     count <- 0
+#'     if (!is.null(req$session$counter)){
+#'       count <- as.numeric(req$session$counter)
+#'     }
+#'     req$session$counter <- count + 1
+#'     return(paste0("This is visit #", count))
+#'   }) %>%
+#'   pr_run()
+#'
+#'
+#' #### -------------------------------- ###
+#'
+#'
+#' ## Save key to a local file
+#' pswd_file <- "normal_file.txt"
+#' cat(plumber::randomCookieKey(), file = pswd_file)
+#' # Make file read-only
+#' Sys.chmod(pswd_file, mode = "0600")
+#'
+#' pr() %>%
+#'   pr_cookie(
+#'     readLines(pswd_file, warn = FALSE),
+#'     name = "counter"
+#'   ) %>%
+#'   pr_get("/sessionCounter", function(req) {
+#'     count <- 0
+#'     if (!is.null(req$session$counter)){
+#'       count <- as.numeric(req$session$counter)
+#'     }
+#'     req$session$counter <- count + 1
+#'     return(paste0("This is visit #", count))
+#'   }) %>%
+#'   pr_run()
+#' }
+#' @export
+pr_cookie <- function(pr,
+                      key,
+                      name = "plumber",
+                      expiration = FALSE,
+                      http = TRUE,
+                      secure = FALSE) {
+  pr$registerHooks(
+    sessionCookie(key = key, name = name, expiration = expiration, http = http, secure = secure)
+  )
+  invisible(pr)
+}
+
+#' Set the default serializer of the router
+#'
+#' By default, Plumber serializes responses to JSON. This function updates the
+#' default serializer to the function supplied via \code{serializer}
+#'
+#' @param pr A plumber router
+#' @param serializer a serializer function
+#'
+#' @return The plumber router with the new default serializer
+#'
+#' @export
+pr_serializer <- function(pr,
+                          serializer) {
+  pr$setSerializer(serializer)
+  invisible(pr)
+}
```

Comment:
Should have some way to distinguish functions that add stuff to `pr` and other functions that just change the value of a default

Reply:
Maybe `pr_modify_*()`?

Reply:
We originally had these as `pr_default_*()` but it felt overly verbose. I'm open to revisiting an alternative naming convention for these though.

Reply:
What about `pr_set_*()`?
cc @schloerke